# The Sparks Foundation #GRIPJUNE21

## Computer Vision & Internet of Things

### Task 2 : COLOUR IDENTIFICATION IN IMAGES.

### By - DIPAN MANDAL

### What is Colour Detection?
 - Colour detection is the process of detecting the name of any color. Simple isn’t it? Well, for humans this is an extremely easy task but for computers, it is not straightforward. Human eyes and brains work together to translate light into color. Light receptors that are present in our eyes transmit the signal to the brain. Our brain then recognizes the color. Since childhood, we have mapped certain lights with their color names. We will be using the somewhat same strategy to detect color names.
 
### About the Python Project
 - In this color detection Python project, we are going to build an application through which you can automatically get the name of the color by clicking on them. So for this, we will have a data file that contains the color name and its values. Then we will calculate the distance from each color and find the shortest one.
 
 

#### Importing the dependencies.

In [6]:
import cv2
import pandas as pd

#### Reading the Image

In [7]:
img_path = r'picture.jpg'    # The path of the Image file.

Img = cv2.imread(img_path)

In [8]:
# Declaring Some Global Variables, will use them later on.
clicked = False
r = g = b = x_pos = y_pos = 0

<b>Here, we have taken a dataset of colours called " colors.csv ", with the help of this dataset we are going to match the colours in the image. So, let's see how many colours it contains. 

In [9]:
# Reading csv file with pandas and giving names to each column
index = ["color", "color_name", "hex", "R", "G", "B"]
csv = pd.read_csv('colors.csv', names=index, header=None)

print('- The Dataset to match from is \n\n',csv)

- The Dataset to match from is 

                     color             color_name      hex    R    G    B
0      air_force_blue_raf   Air Force Blue (Raf)  #5d8aa8   93  138  168
1     air_force_blue_usaf  Air Force Blue (Usaf)  #00308f    0   48  143
2    air_superiority_blue   Air Superiority Blue  #72a0c1  114  160  193
3         alabama_crimson        Alabama Crimson  #a32638  163   38   56
4              alice_blue             Alice Blue  #f0f8ff  240  248  255
..                    ...                    ...      ...  ...  ...  ...
860         yellow_orange          Yellow Orange  #ffae42  255  174   66
861        yellow_process       Yellow (Process)  #ffef00  255  239    0
862            yellow_ryb           Yellow (Ryb)  #fefe33  254  254   51
863                zaffre                 Zaffre  #0014a8    0   20  168
864     zinnwaldite_brown      Zinnwaldite Brown  #2c1608   44   22    8

[865 rows x 6 columns]


### So we saw that the dataset contains 865 different colours, and their hexadecimal values along with their R,G,B values.
So, now we will be creating two functions named, **"get_color_name( )"** and **"draw_function"** . 

- The **draw_function** will be giving us the R, G, B values of the cursor position where we will **double left click**. (Any other sort of click can also be used in place of double left click). The **R, G, B** values are then stored into **global variables** for further use.


- Another function is **get_color_name**, this function takes the R, G, B values from the draw_function and matches it with the dataset of colours(colors.csv) and returns the name of the closest colour it found while matching.

In [10]:
# function to calculate minimum distance from all colors and get the most matching color
def get_color_name(R, G, B):
    minimum = 10000
    for i in range(len(csv)):
        d = abs(R - int(csv.loc[i, "R"])) + abs(G - int(csv.loc[i, "G"])) + abs(B - int(csv.loc[i, "B"]))
        if (d <= minimum):
            minimum = d
            cname = csv.loc[i, "color_name"]
    return cname

# function to get x,y coordinates of mouse double click
def draw_function(event, x, y, flags, param):
    if event == cv2.EVENT_LBUTTONDBLCLK:
        global b, g, r, x_pos, y_pos, clicked
        clicked = True
        x_pos = x
        y_pos = y
        b, g, r = Img[y, x]
        b = int(b)
        g = int(g)
        r = int(r)


cv2.namedWindow('Image')
cv2.setMouseCallback('Image', draw_function)

while True:

    cv2.imshow("Image", Img)
    if clicked:

        # cv2.rectangle(image, start point, endpoint, color, thickness), -1 fills entire rectangle
        cv2.rectangle(Img, (20, 20), (750, 60), (b, g, r), -1)

        # Creating text string to display( Color name and RGB values )
        text = get_color_name(r, g, b) + ' R=' + str(r) + ' G=' + str(g) + ' B=' + str(b)

        # cv2.putText(img,text,start,font(0-7),fontScale,color,thickness,lineType )
        cv2.putText(Img, text, (50, 50), 2, 0.8, (255, 255, 255), 2, cv2.LINE_AA)

        # For very light colours we will display text in black colour
        if r + g + b >= 600:
            cv2.putText(Img, text, (50, 50), 2, 0.8, (0, 0, 0), 2, cv2.LINE_AA)

        clicked = False

    # Break the loop when user hits 'esc' key
    if cv2.waitKey(20) & 0xFF == 27:
        break

cv2.destroyAllWindows()


## This concludes the topic, Color Identification using an Image